File: factorypattern/factorypattern.js
// Abstract Shape class
class Shape {
  draw() {
    throw new Error('Abstract method "draw()" must be implemented');
  }
}

// Rectangle class
class Rectangle extends Shape {
  draw() {
    console.log('It is Rectangle');
  }
}

// Circle class
class Circle extends Shape {
  draw() {
    console.log('It is Circle');
  }
}

// Square class
class Square extends Shape {
  draw() {
    console.log('It is Square');
  }
}

// ShapeFactory class
class ShapeFactory {
  getShape(input) {
    switch (input.toUpperCase()) {
      case 'CIRCLE':
        return new Circle();
      case 'RECTANGLE':
        return new Rectangle();
      default:
        return new Square();
    }
  }
}

// Usage
const shapeFactoryObj = new ShapeFactory();

const shapeObj1 = shapeFactoryObj.getShape('CIRCLE');   // Circle object
const shapeObj2 = shapeFactoryObj.getShape('SQUARE');   // Square object

shapeObj1.draw();
shapeObj2.draw();

File: factorypattern/factorypattern.py
from abc import ABC, abstractmethod

class Shape(ABC):
	@abstractmethod
	def draw(self):
		pass

class Rectangle(Shape):
	def draw(self):
		print('It is Rectangle')

class Circle(Shape):
	def draw(self):
		print('It is Circle')

class Square(Shape):
	def draw(self):
		print('It is Square')

class ShapeFactory():
	def get_shape(self, input):
		if input == 'CIRCLE':
			return Circle()
		elif input == 'RECTANGLE':
			return Rectangle()
		else:
			return Square()

if __name__ == '__main__':
	shape_factory_obj = ShapeFactory()
	shape_obj_1 = shape_factory_obj.get_shape('CIRCLE')	# circle object
	shape_obj_2 = shape_factory_obj.get_shape('SQUARE')	# square object

	shape_obj_1.draw()
	shape_obj_2.draw()

